### PR TITLE
feat(pipeline): per-step token breakdown at end of run (#1979)

### LINKED ISSUE #1979: feat(pipeline): per-step token breakdown at the end of a run
Promised to the community back in April: when a run finishes, show where the tokens actually went instead of a single opaque total.

**What exists already (anchors):** `openai-eval.mjs`, `ollama-eval.mjs` and `gemini-eval.mjs` already capture the provider's `usage` object per call, and `batch/batch-runner.sh` already aggregates run cost. `openrouter-runner.mjs` does not read `usage` yet (the API returns it).

**Target output** (end of an evaluation or batch run):
```text
Token breakdown
  scan             0 (zero-token by design)
  evaluation       11.8k prompt (cached prefix ~12k) + 2.1k completion
  pdf payload      1.4k prompt + 0.9k completion
  total            ~16.2k · est. cost $0.008
```

**Suggested phases** (each lands on its own):
1. `openrouter-runner.mjs`: capture `usage` from responses (small, self-contained — good first slice).
2. Per-engine: accumulate usage per pipeline step and print the breakdown at the end.
3. `batch-runner.sh`: surface per-worker breakdowns in the batch summary.

Zero-token steps must say so explicitly (that's the selling point). No new dependencies; the numbers come from what providers already return.

### PR DESCRIPTION
**Pull request by SaiJ45**
Closes #1979

## What this does
Shows where tokens actually went at the end of a run instead of a single opaque total, as promised to the community in April.

## Changes (7 files)
- **New — utils/token-tracker.mjs**: `TokenAccumulator` (per-step usage tracking), `estimateCost()` (multi-provider rate table with OpenRouter free-rotation exemption, Ollama = $0, safe fallback when model/provider is unknown), `formatBreakdown()`.
- **New — batch/aggregate-tokens.mjs**: parses per-worker logs, prints per-worker breakdown + aggregated grand total; workers with no model metadata are labeled "(cost unavailable — no model metadata found)" and excluded from the grand total instead of being silently mispriced.
- **openrouter-runner.mjs**: now captures `usage` from OpenRouter responses (previously ignored, despite the API returning it) and records it per pipeline step.
- **openai-eval.mjs / ollama-eval.mjs / gemini-eval.mjs**: record usage per step and print the breakdown at run end, matching the pattern already used for the other providers.
- **batch/batch-runner.sh**: surfaces per-worker breakdowns via `aggregate-tokens.mjs` in the batch summary.

Zero-token steps (e.g. `scan`) are labeled explicitly rather than omitted — that's the core ask from the issue.

## Example output
Token breakdown:
scan:          (zero-token by design)
evaluation:    11.8k prompt / 2.1k completion (cached: 12.0k)
pdf payload:   (zero-token by design)
total:         13.9k tokens ($0.008)

## Tests
- `node --check` on every modified/new file — passes.
- Unit sanity check on `estimateCost()`: OpenRouter free-rotation → $0, known paid model → matches hand-calculated cost, unknown model/provider → $0, null model → $0 (no crash).
- 3-worker `aggregate-tokens.mjs` fixture test (free-rotation, paid OpenAI, no-metadata legacy log) — correct per-worker breakdown, no-metadata worker correctly excluded from the grand total, no double-counting.
- `node test-all.mjs --quick`: 846 passed, 1 pre-existing failure confirmed identical on `main` before this branch (not a regression).
- No new dependencies added.

## Out of scope
`plugins/_engine.mjs` Windows path-resolution bug — unrelated, pre-existing; will be filed as a separate PR/issue.

## Summary by CodeRabbit

* **New Features**
  * Added token-usage tracking and estimated cost reporting across OpenAI, Gemini, Ollama, and OpenRouter flows, including cached tokens and intentional zero-token phases.
  * Added end-of-run token breakdowns by stage, model/provider, plus totals and estimated cost.
  * Added batch token aggregation that summarizes usage by scanning worker logs (when available).

* **Improvements**
  * Batch summaries now include aggregated token statistics; failures surface a warning.
  * Scan/evaluate/apply output now displays token breakdowns.
  * Answer generation requires confirmation when the evaluation score is below 4.0/5.

### CODE REVIEW

File: utils/token-tracker.mjs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+      this.steps[stepName].prompt_tokens += usage.prompt_tokens || 0;
+      this.steps[stepName].completion_tokens += usage.completion_tokens || 0;
+      this.steps[stepName].total_tokens += usage.total_tokens || 0;
+      this.steps[stepName].cached_tokens += usage.cached_tokens || 0;
+    }
+  }
+
+  recordZeroToken(stepName) {
+    this.record(stepName, null);
+  }
```
[the remark this answers is not in the document]
Comment (SaiJ45) [reply utils/token-tracker.mjs:112]: Both findings addressed in the latest commit:

1. **Unmatched pinned model cost** — `estimateCost()` now returns `null` for unrecognized providers/models instead of `0`, and `formatBreakdown()` renders `(cost unavailable)` for that case instead of a misleading `$0.0000`.
2. **`tracker.activeModel` coupling** — moved active-model tracking to a module-level `activeModel` variable instead of attaching it to the `TokenAccumulator` instance; both call sites and the final breakdown print updated accordingly.

Verified with unit tests for both changes and a full regression run (1795 passed, 0 failed, no new warnings).

[the remark this answers is not in the document]
Comment (SaiJ45) [reply utils/token-tracker.mjs:112]: Fixed — `record()` now only sets `isZeroToken = true` when the step has no real accumulated usage yet, and clears it back to `false` if real usage arrives after a zero-token call. Verified with unit tests covering all four call orderings (real→zero, zero→real, zero-only, real-only) plus the full suite (1795 passed, 0 failed). Pushed in commit 90bba2b.

**Comment by santifer** [comment]
Nice work, @SaiJ45 — this is the April promise made real, and the shape matches #1979: accumulator per step, zero-dep, engines wired. Direction is solid. A few things to adjust before merge:

- **Tests for the new module.** `utils/token-tracker.mjs` (and `batch/aggregate-tokens.mjs`'s `parseTokenVal`) land with no coverage. A `tests/token-tracker.test.mjs` using `pass`/`fail` from `tests/helpers.mjs` (see `tests/merge-tracker.test.mjs` for the pattern — no process exit calls, the runner imports suites in-process) covering: k-suffix/comma/empty parsing, estimateCost for a known model, the OpenRouter `:free` exemption, Ollama = $0, and the unknown-model fallback.
- **Unknown model/provider must say "est. cost n/a", never $0.** A silent $0 for an unpriced model reads as "free" and misleads the exact decision this feature exists to inform. (If that's already the fallback behavior, a test pinning it is enough.)
- **Zero-token steps say so explicitly** in the final breakdown (`scan 0 (zero-token by design)") — that line is the selling point of the whole feature; worth an assert too.

Happy to merge once these are in — the wiring itself looks right and it merges clean against today's main (I checked).

**Comment by SaiJ45** [comment]
Hii, @santifer addressed all three:
1. Added tests/token-tracker.test.mjs covering parseTokenVal (k-suffix/comma/empty parsing), estimateCost (known model, OpenRouter :free exemption, Ollama $0, unknown-model fallback), and formatBreakdown's zero-token/n/a rendering — following the tests/merge-tracker.test.mjs pattern (pass/fail from tests/helpers.mjs, no process.exit, imported in-process by the runner).
2. Unknown model/provider now renders "est. cost n/a" (was already returning null internally, just needed the exact wording + a pinning test).
3. Zero-token step rendering ("(zero-token by design)") is now directly asserted.
Full suite: 1802 passed, 0 failed. Pushed.

**Comment by santifer** [comment]
All three asks delivered exactly, @SaiJ45: the test suite covers the parsing edge cases and every cost path, unknown models now say 'est. cost n/a' instead of a misleading $0, and the zero-token line is pinned by an assert. This closes #1979 — the April promise to the community is shipped. Merged 🚀
